src/BucketTesting/BucketSelector.php

```diff
@@ -27,14 +29,28 @@ private function sanitizeParameters( array $params ): array {
 	public function selectBuckets( array $cookie = [], array $urlParameters = [] ): array {
 		$urlParameters = $this->sanitizeParameters( $urlParameters );
 		$cookie = $this->sanitizeParameters( $cookie );
-		$possibleKeys = array_merge( $cookie, $urlParameters );
+		$possibleParameters = array_merge( $cookie, $urlParameters );
 
-		[ $buckets, $missingCampaigns ] = $this->campaigns->splitBucketsFromCampaigns( $possibleKeys );
+		$selectionStrategies = [
+			new InactiveCampaignBucketSelection(),
+			new ParameterBucketSelection( $possibleParameters ),
+			$this->fallbackSelectionStrategy
+		];
 
-		foreach( $missingCampaigns as $campaign ) {
-			$buckets[] = $this->selectionStrategy->selectBucketFromCampaign( $campaign );
+		$buckets = [];
+		foreach( $this->campaigns as $campaign ) {
+			$buckets[] = $this->selectBucketWithStrategies( $selectionStrategies, $campaign );
 		}
 		return $buckets;
 	}
 
+	private function selectBucketWithStrategies( array $selectionStrategies, Campaign $campaign ): Bucket {
+		foreach( $selectionStrategies as $strategy ) {
+			if ( $bucket = $strategy->selectBucketForCampaign( $campaign ) ) {
+				return $bucket;
+			}
+		}
+		throw new InvalidArgumentException( 'Fallback bucket selection must always return a bucket.' );
```

Comment:
Maybe https://secure.php.net/manual/en/class.outofboundsexception.php is more appropriate

Reply:
This exception can only be caused if the `fallbackSelectionStrategy` argument passed into the `BucketSelector` constructor does not return a bucket. Not because some out of bounds situation. I see that it can be a bit confusing because the argument was not passed directly to the method, but to the constructor instead. I'll throw a `LogicException` instead.